src/mon/Monitor.cc

```diff
@@ -2006,7 +2006,12 @@ void Monitor::handle_probe_reply(MonOpRequestRef op)
 	       << ", mine was " << monmap->get_epoch() << dendl;
       delete newmap;
       monmap->decode(m->monmap_bl);
-      notify_new_monmap(false);
+      if (!has_ever_joined) {
+        dout(20) << " !has_ever_joined " << *monmap << dendl;
+        notify_new_monmap(false, false);
```

Comment:
Hmm. I forget when we start pinging and generating ConnectionTracker state. Is the following possible?

New monitor turns on
Start pinging existing 5 monitors
Start copying paxos state from mon.1
mon.4 is removed from monitor list
finish copying state
get new map with mon.4 in removed_ranks
fail to remove because we are !has_ever_joined

Reply:
Hi Greg,

How about we only allow mon remove when it has joined quorum before, for other cases like
never joined or joined but we have a very outdated version, we don't adjust and also make
sure to nuke peer_report, incase it has already been populated with like a removed_rank:

````
      if (newmap->get_epoch() - monmap->get_epoch() == 1 &&
              has_ever_joined) {
        notify_new_monmap(false);
      } else {
        notify_new_monmap(false, false);
        elector.notify_clear_peer_state();
      }
````

Reply:
I don't think we can block mon removal on stuff like that — what if there's physical hardware loss between adding it to the map and when it actually comes on?
And that wouldn't fix the scenario I proposed either — the problem is just having a new monitor see a different removed monitor at a bad time.

Reply:
by state, do you mean `peer_tracker` in the elector? if so I guess you mean even if we are 1 epoch apart we should always `elector.notify_clear_peer_state();`

Reply:
Right, if we see a discontinuity we need to reset our scores because we don't know what might have happened to the monitor set and how the scores should align.